### DISCUSSION TITLE
Pre wakeup

### DISCUSSION THREAD

**Discussion by lonesomebyte537** [zephyrproject-rtos/zephyr#53250] [Q&A]
Hi, I have an interrupt in my system that needs immediate servicing. By assigning the right priority, it works as expected. But with power management enabled, the exit latency of the low power mode delays the isr too much. The earliest time the interrupt will happen can be "predicted". So I was thinking of following possible solutions: 

- Somehow modify the timeout value that is passed by the power manager to the timer driver such that the system wakes up at the earliest point
- Use a dummy thread or delayable work item to wake up the system in time
- Use a hardware timer to generate another interrupt prior to the time critical interrupt

The first solution is the preferred one as this does not require additional resources but not sure whether this can be achieved without having to modify the power manager.
Anyone other suggestions on how to solve this?

Thanks!

**Comment by yashi** [comment]
> not sure whether this can be achieved without having to modify the power manager.

Does your system have a way to come up from a low power state without any external events?  How does it work? If there is a way, you can write it in `pm_policy_next_state()`.

Here is more general answer.

Are you saying that, your system
- has periodical or deterministic interrupts from your device(s).
- has a low power mode and CPU needs some extra time to recover from it.

And you want to
- enable low power mode to save power.
- but, you want to serve your devices as quickly or as constantly as possible?

I don't know how precise your requirements, hardware interrupts, or system timers are, but I'd
- use a system or hardware timer.
- access your devices when device interrupts the CPU.

This assumes that
- The current PM API does _not_ have "the timeout", which I assume that you mean "the duration to stay in the low power state".
- Your system or hardware power management system needs an extra event to wake up.

Hope this helps.
